### PR TITLE
Better installer logs

### LINKED ISSUE #1444: Mac: 'gvfs upgrade' reports failures and logging is insufficient to diagnose
Two issues:

- The upgrader is reporting errors (even though the upgrade appears to have succeeded)
- The logging is insufficient to determine what happened

User facing error:

```
sudo gvfs upgrade --confirm --no-verify
Checking for GVFS upgrades...Succeeded
New version 0.6.19224.2 is available.
Launching upgrade tool...
Downloading...Succeeded
Unmounting repositories...Succeeded
Checking for blocking processes....Succeeded
Running Git Installer...Succeeded
Running GCM Core Installer...Succeeded
Running Configure GCM Script...Succeeded
Running Git Service Daemon Installer...Succeeded
Running VFS for Git Installer...Failed
 
ERROR: The VFS for Git Installer failed, but no error message was provided by the failing command.
```

Logs:

```
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: -[IFDInstallController(Private) _buildInstallPlanReturningError:]: file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Set authorization level to root for session
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Administrator authorization granted.
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Will use PK session
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Using authorization level of root for IFPKInstallElement
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Starting installation:
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Configuring volume "Sierra"
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Preparing disk for local booted install.
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Free space on "Sierra": 25.5 GB (25497071616 bytes).
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: Create temporary directory "/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T//Install.530973fZlxL"
2019-08-12 16:52:39-07 <Machine Name> installer[53097]: IFPKInstallElement (1 packages)
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Adding client PKInstallDaemonClient pid=53097, uid=0 (/usr/sbin/installer)
2019-08-12 16:52:40-07 <Machine Name> installer[53097]: PackageKit: Enqueuing install with framework-specified quality of service (utility)
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: ----- Begin install -----
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: request=PKInstallRequest <1 packages, destination=/>
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: packages=(
                    "PKLeopardPackage <id=com.vfsforgit.pkg, version=0, url=file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg>"
                )
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Set reponsibility for install to 1191
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Extracting file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg (destination=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/C/PKInstallSandboxManager/EC99C483-0009-4283-88BB-11C8C15D4A0C.activeSandbox/Root, uid=0)
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: prevent user idle system sleep
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: suspending backupd
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Executing script "./preinstall" in /private/tmp/PKInstallSandbox.G9D4rX/Scripts/com.vfsforgit.pkg.vacAkD
2019-08-12 16:52:40-07 <Machine Name> install_monitor[53099]: Temporarily excluding: /Applications, /Library, /System, /bin, /private, /sbin, /usr
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: ./preinstall: /sbin/kextunload -b org.vfsforgit.PrjFSKext
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Using trashcan path /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/PKInstallSandboxTrash/EC99C483-0009-4283-88BB-11C8C15D4A0C.sandboxTrash for sandbox /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/C/PKInstallSandboxManager/EC99C483-0009-4283-88BB-11C8C15D4A0C.activeSandbox
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Shoving /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/C/PKInstallSandboxManager/EC99C483-0009-4283-88BB-11C8C15D4A0C.activeSandbox/Root (2 items) to /
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: PackageKit: Executing script "./postinstall" in /private/tmp/PKInstallSandbox.G9D4rX/Scripts/com.vfsforgit.pkg.vacAkD
2019-08-12 16:52:40-07 <Machine Name> installd[6157]: ./postinstall: /sbin/kextload "/Library/Extensions/PrjFSKext.kext"
2019-08-12 16:52:41-07 <Machine Name> installd[6157]: ./postinstall: Restarting Service: '/bin/launchctl kickstart -k system/org.vfsforgit.prjfs.PrjFSKextLogDaemon'
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: ./postinstall: Restarting Service: '/bin/launchctl kickstart -k gui/501/org.vfsforgit.usernotification'
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: ./postinstall: Restarting Service: '/bin/launchctl kickstart -k gui/501/org.vfsforgit.service'
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: ./postinstall: Could not find domain for
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: ./postinstall: Starting Service: '/bin/launchctl bootstrap gui/0 /Library/LaunchAgents/org.vfsforgit.usernotification.plist'
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: ./postinstall: Could not find domain for
2019-08-12 16:52:52-07 <Machine Name> install_monitor[53099]: Re-included: /Applications, /Library, /System, /bin, /private, /sbin, /usr
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: releasing backupd
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: allow user idle system sleep
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: Install Failed: Error Domain=PKInstallErrorDomain Code=112 "An error occurred while running scripts from the package “VFSForGit.0.6.19224.2.pkg”." UserInfo={NSFilePath=./postinstall, NSURL=file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg, PKInstallPackageIdentifier=com.vfsforgit.pkg, NSLocalizedDescription=An error occurred while running scripts from the package “VFSForGit.0.6.19224.2.pkg”.} {
                    NSFilePath = "./postinstall";
                    NSLocalizedDescription = "An error occurred while running scripts from the package \U201cVFSForGit.0.6.19224.2.pkg\U201d.";
                    NSURL = "file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg";
                    PKInstallPackageIdentifier = "com.vfsforgit.pkg";
                }
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: Running idle tasks
2019-08-12 16:52:52-07 <Machine Name> installer[53097]: install:didFailWithError:Error Domain=PKInstallErrorDomain Code=112 "An error occurred while running scripts from the package “VFSForGit.0.6.19224.2.pkg”." UserInfo={NSFilePath=./postinstall, NSURL=file://localhost/usr/local/vfsforgit_upgrader/Downloads/InstallerTemp/content/Installers/MacOS/VFSForGit/VFSForGit.0.6.19224.2.pkg, PKInstallPackageIdentifier=com.vfsforgit.pkg, NSLocalizedDescription=An error occurred while running scripts from the package “VFSForGit.0.6.19224.2.pkg”.}
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: Removing client PKInstallDaemonClient pid=53097, uid=0 (/usr/sbin/installer)
2019-08-12 16:52:52-07 <Machine Name> installd[6157]: PackageKit: Done with sandbox removals
2019-08-12 16:52:53-07 <Machine Name> installer[53097]: Install failed: The Installer encountered an error that caused the installation to fail. Contact the software manufacturer for assistance.
```

### PR DESCRIPTION
**Pull request by alameenshah**
#### Fix mac installer/upgrader failure

VFSForGit installer fails in postinstall script while trying to launch Service and UserNotification agents in GUI context of logged in users. To determine users logged into the GUI context, postinstall script looks for loginwindow commands and the user running the command. When Fast User switching is enabled, then loginwindow can be running as root user, even though root user is Not logged in. In such cases, postinstall tries to launch Service and UserNotifcation agents in GUI context of root user (that does not exist) resulting in failure. This commit makes
postinstall look for Finder command (which only runs after a user has successfully logged into GUI context and is therefore more reliably associated with GUI session) rather than
loginwindow.

Also Updated launchd plist for Service and Notification agents, added LimitLoadToSessionType = Aqua, just to make sure the agents get launched into the GUI context only.

#### [macOS] Capture installer log on upgrade failure
- Save last 100KB of /var/log/installer.log immediately after  an upgrade failure.
- Installer logs would be available through `gvfs diagnose`.
- Added new platform property SupportsSystemInstallLog. This will be  set to true for Platforms(macOS) that keep system wide installer log (vs custom logs specific to VFSForGit installer package).

#### Steps to repro mac installer/upgrader failure issue - 
1. Click on "Login Window..." item from "FastUser Switching" menu.
Screen Shot 2019-08-28 at 1 49 46 PM

2. A login window prompting for user credentials would appear. Login as the same user you were logged-in as in step 1.
3. Now there will be a `loginwindow` process running as `root` - even though no `root` user is logged in.
Screen Shot 2019-08-28 at 1 52 52 PM

4. Now run `VFSForGit` installer manually or run `sudo gvfs upgrade --confirm --no-verify` and notice that the installation fails in the `postinstall` script, while trying to auto-launch `VFSForGit` services for `root` user.

#### References 
- [ Login sessions](https://developer.apple.com/library/archive/technotes/tn2083/_index.html#//apple_ref/doc/uid/DTS10003794-CH1-SUBSECTION3)

Fixes #1444

### CODE REVIEW

File: GVFS/GVFS.Installer.Mac/scripts/postinstall [hunk 1]
```diff
@@ -1,24 +1,39 @@
...
-	restartCmd="/bin/launchctl kickstart -k $domain/$service"
-	isLoaded=`/bin/launchctl print $domain/$service | wc -l`
-	if [ $isLoaded -gt "0" ]; then
-		echo "Restarting Service: '$restartCmd'"
-		eval $restartCmd || exit 1
-	else
-		echo "Starting Service: '$startCmd'"
-		eval $startCmd || exit 1
-	fi
+    domain=$1
```
Comment (alameenshah) [inline GVFS/GVFS.Installer.Mac/scripts/postinstall]: replaced tabs with spaces

File: GVFS/GVFS.Notifications/VFSForGit.Mac/org.vfsforgit.usernotification.plist [hunk 2]
```diff
@@ -24,6 +24,8 @@
         <string>/usr/local/vfsforgit</string>
         <key>ProcessType</key>
         <string>Interactive</string>
+        <key>LimitLoadToSessionType</key>
```
Comment (alameenshah) [inline GVFS/GVFS.Notifications/VFSForGit.Mac/org.vfsforgit.usernotification.plist]: Find more info on various session types [here](https://developer.apple.com/library/archive/technotes/tn2083/_index.html#//apple_ref/doc/uid/DTS10003794-CH1-SUBSECTION3)

File: GVFS/GVFS.Installer.Mac/scripts/postinstall [hunk 3]
```diff
@@ -1,24 +1,39 @@
 #!/bin/bash
 
+LOGFILENAME=`/bin/date +$HOME/Library/Application\ Support/GVFS/GVFS.Upgrade/UpgraderLogs/gvfs_installer_postinstall_%Y%m%d_%H%M%S.log`
```
Comment (nickgra) [inline GVFS/GVFS.Installer.Mac/scripts/postinstall]: Does this work fine if we don't already have the `GVFS/GVFS.Upgrade` folder in /Library/Application Support?
What's the permissions on this folder if it's created by the postinstall script? Will it be owned by root and therefore un-writable by other parts of the installer/upgrader process?

Comment (jamill) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: good question!

Can we control the location of the logs from the command line? When running upgrade, we might want the logs to go into the "upgrader" directory. But if we are just running the installer directly, we might not want the logs there... On the Windows side of things, the upgrader app passes in the log directory. Can we do something similar here?

This also hardcodes the location of the installer logs here to match what is in code - so if we ever update the log directory, it we will need to remember that there are 2 locations to update.

Comment (alameenshah) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: >> Does this work fine if we don't already have the GVFS/GVFS.Upgrade folder in /Library/Application Support?

No it doesn't - fresh installs would have failed (in-fact [one of our pipeline builds](https://dev.azure.com/mseng/AzureDevOps/_build/results?buildId=10232664&view=results) were failing for this reason). Added [new commit](https://github.com/microsoft/VFSForGit/pull/1466/commits/12f29c829b716713d91b64d788ba4f86056799cd) that creates the directory if it does not exist.

>> What's the permissions on this folder if it's created by the postinstall script? Will it be owned by root and therefore un-writable by other parts of the installer/upgrader process?

[This commit](https://github.com/microsoft/VFSForGit/pull/1466/commits/12f29c829b716713d91b64d788ba4f86056799cd) creates the directory using `sudo -u $USER` where `USER` is an installer variable that refers to the active logged-in user (and not `root` user) - the directory gets created with the ownership and permissions of  `USER`, even when running as `root` during `sudo gvfs upgrade --confirm`.

This is how it looks on my Mac
```
$ ls -Rl ~/Library/Application\ Support/GVFS/
total 0
drwxr-xr-x  4 Ameen  staff  128 Aug 21 10:39 GVFS.Upgrade

/Users/Ameen/Library/Application Support/GVFS//GVFS.Upgrade:
total 0
drwxr-xr-x  4 Ameen  staff  128 Aug 21 10:40 UpgraderLogs
```
>>Can we control the location of the logs from the command line?

Installer does not seem to provide an option to pass custom args from command line to installer scripts.

>> This also hardcodes the location of the installer logs here to match what is in code - so if we ever update the log directory, it we will need to remember that there are 2 locations to update.

Agree.

Comment (alameenshah) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: @jrbriggs - thoughts?

Comment (jrbriggs) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: Where would stdout go if we're not teeing to the log file? Is there an opportunity to play closer with the OS and not need to take responsibility for creating these files?

If not, this seems like the best compromise. Thanks!

Comment (jamill) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: My question was more around the fact that it seemed odd that the installer would write its logs to the "upgrade" folder, even if the installation was happening outside of an `gvfs upgrade` workflow. It now creates implicit dependencies / assumptions between the `gvfs upgrade` code and the installer package.

It is a convenient way when the installer is being run as part of the upgrade, so maybe it is a good compromise.

Comment (wilbaker) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: > It is a convenient way when the installer is being run as part of the upgrade, so maybe it is a good compromise.

Would it be possible and/or make sense to write the installer logs to `GVFS.Installer` folder (regardless of whether or not the installer is running as part of an upgrade)?  I'm not sure if that would be less confusing?  Thoughts?

Comment (alameenshah) [reply GVFS/GVFS.Installer.Mac/scripts/postinstall]: >> Where would stdout go if we're not teeing to the log file? Is there an opportunity to play closer with the OS and not need to take responsibility for creating these files?

This is possible if we capture `/var/log/install.log`. It holds error logs from `vfsforgit`'s installer scripts and also failure info from `git`, `gcm-core` installers. Compared to custom installer logs from installer scripts, this is non-lossy. It gets created only during  failures in `gvfs upgrade` (and not during manual installs using GUI.) 

Added [commit](https://github.com/microsoft/VFSForGit/pull/1466/commits/b2d69ec1c23429bec609756a9acc30c5529cd77e).

**Pull request merged**